modules/glue/iam.tf

```diff
@@ -23,37 +27,40 @@ EOF
   tags               = var.tags
 }
 
+data "aws_iam_policy_document" "retraining_glue" {
+  statement {
+    sid = "S3Access"
+    actions = [
+      "s3:ListBucket"
+    ]
+    resources = [for bucket in local.s3_buckets : "arn:aws:s3:::${bucket}"]
+  }
 
-resource "aws_iam_policy" "retraining_glue_policy" {
-  name   = "${var.model_name}-retraining-glue-policy"
-  policy = <<EOT
-{
-    "Version": "2012-10-17",
-    "Statement": [
-        {
-            "Sid": "S3BucketsAccess",
-            "Effect": "Allow",
-            "Action": [
-                "s3:GetObject",
-                "s3:PutObject"
-            ],
-            "Resource": [
-                "arn:aws:s3:::${var.data_location_s3}/*", 
-                "arn:aws:s3:::${var.config_bucket_id}/*"
-            ]
-        },
-        {
-          "Sid": "AllowAccessToKey",
-          "Effect": "Allow",
-          "Action": [
-            "kms:Decrypt", 
-            "kms:GenerateDataKey"
-          ],
-          "Resource": "arn:aws:kms:${var.region}:${var.account_id}:key/*"
-        }
+  statement {
+    sid = "S3ObjectAccess"
+    actions = [
+      "s3:GetObject",
+      "s3:PutObject",
+      "s3:DeleteObject",
     ]
+
+    resources = [for bucket in local.s3_buckets : "arn:aws:s3:::${bucket}/*"]
+  }
+
+  statement {
+    sid = "KMSAccess"
+    actions = [
+      "kms:Decrypt",
+      "kms:GenerateDataKey"
+    ]
+
+    resources = local.kms_key_arns
```

Comment:
Only giving permissions to the keys and buckets required. Giving permission to all keys in the accounts is potentially insecure (principle of least privilege)